glusterfs: Allow admin to enable or disable custom endpoint/service names

humblec opened the pull request:
This patch introduces a new SC parameter called `customepnameprefix`
    in glusterfs plugin. This new storageclass parameter allow admins
    to create endpoints and services with mentioned value of this
    parameter. If this parameter has not been set or if its empty,
    default prefix string ie "glusterfs-dynamic-" will be used for ep/svc
    name.
        
This patch address below issues#
https://github.com/kubernetes/kubernetes/issues/53939
https://github.com/kubernetes/kubernetes/issues/56379

Additional Ref# https://github.com/kubernetes/kubernetes/pull/56380

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

humblec commented:
/release-note-none

humblec commented:
/test pull-kubernetes-integration

humblec commented:
/assign @jsafrane

humblec commented:
/retest

humblec commented:
@jsafrane PTAL . :)

humblec commented:
> And if you use PVC uuid as base of the name, it has guaranteed length and "glusterfs-dynamic-" prefix will always fit.>

Sure, last commit changed to UID as per review comment. The small concern here is the backward compatibility of ep name, but we still have `glusterfs-dynamic-` string in it. so , I think its fine. More or less, its internal to provisioner and admin should not completely relay on it.  Thanks!

humblec commented:
@obnoxxx the latest commit addressing Jan's review comment will actually solve the length issue.

jsafrane commented:
If you use PVC.UID as base of the name, is there any reason to make `customepname` configurable? It's introduced to solve the name length issue, but you don't have it now.

> The small concern here is the backward compatibility of ep name, but we still have glusterfs-dynamic- string in it. so , I think its fine. 

If I have a running cluster with couple of provisioned volumes and update to the new version, I will still have endpoints like `glusterfs-dynamic-<pvc.name>`. Only newly provisioned volumes will get `glusterfs-dynamic-<pvc.uid>`. So when you delete endpoints, you should try to delete both of them.

obnoxxx commented:
@humblec right, we can use the uuid. But this should IMHO be a separate PR, because that change is not part of the original topic of this PR. One possible concern is that it's easier to make sense of the endpoint names if they carry the chose name instead of the UUID. I suggest to only use the UUID if customepname == false (or else introduce a second option for controlling the use of the UUID).

obnoxxx commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -815,15 +815,15 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsVolum
 
 	if p.provisionerConfig.customEpName {
-		epServiceName = dynamicEpSvcPrefix + p.options.PVC.Name
+		epServiceName = dynamicEpSvcPrefix + string(p.options.PVC.UID)
```
I think, UUID should only be used if `customEpName == false`.

obnoxxx reviewed (changes requested):
The first patch introducing customepname looks clean except for the cosmetic remarks about unrelated whitespace changes. I think such unrelated changes should not be sneaked in since they unnecessarily bloat the diff of the patch.

obnoxxx commented:
@jsafrane writes:
> If you use PVC.UID as base of the name, is there any reason to make customepname configurable? It's introduced to solve the name length issue, but you don't have it now.

So, we have a few options here in principle:

1. **Always** use the UID for new creates and keep the prefix: `glusterfs-dynamic-<pvc.uid>`.
1. **Always** use the UID for new creates but don't use the prefix: `<pcv.uid>`.
1. Keep the old scheme `glusterfs-dynamic-<pvc.name>` by default but use the UUID (with or without prefix) only if a new config option is set, e.g. `customEpName = false` or `uidAsEpName = true` or so ...
1. Always use the UID, but configure the use of the prefix with the config setting (currently `customEpName`, but maybe better named `useEpPrefix` or so).

I think switching to UID and having the means of disabling or enabling the prefix as in the current  proposal is the worst solution. Which one do you guys prefer, @humblec @jsafrane ?

> > The small concern here is the backward compatibility of ep name, but we still have glusterfs-dynamic- string in it. so , I think its fine.
> 
> If I have a running cluster with couple of provisioned volumes and update to the new version, I will still have endpoints like `glusterfs-dynamic-<pvc.name>`.  Only newly provisioned volumes will get `glusterfs-dynamic-<pvc.uid>`. So when you delete endpoints, you should try to delete both of them.

Excellent point. This needs to be added whichever path we choose.

humblec commented:
> If you use PVC.UID as base of the name, is there any reason to make `customepname` configurable? It's introduced to solve the name length issue, but you don't have it now.
>

Actually thats not the only reason. As the custom naming is  an internal logic, I thought of providing an option to user to enable/disable.
 
> If I have a running cluster with couple of provisioned volumes and update to the new version, I will still have endpoints like `glusterfs-dynamic-<pvc.name>`. Only newly provisioned volumes will get `glusterfs-dynamic-<pvc.uid>`. So when you delete endpoints, you should try to delete both of them.
>

@jsafrane I thought about this before sending last commit and I didnt see a reason to have an extra logic here. :) .  The reason being, at time of delete, we pick the ep from spec. so,  whether its `uuid/name` based doesnt really matter. Or did I miss any other scenarios which can cause the inconsistency ?

obnoxxx commented:
> > If you use PVC.UID as base of the name, is there any reason to make customepname configurable? It's introduced to solve the name length issue, but you don't have it now.
> 
> Actually thats not the only reason. As the custom naming is an internal logic, I thought of providing an option to user to enable/disable.

Thinking about it, the name `customEpName` is confusing, because you do not allow configuration of a full name (or even a custom name), but you enable or disable the use of a (hardcoded) prefix for the endpoint name.... So `enableEpNamePrefix` would sound more appropriate, though somewhat awkward -- maybe enableEpPrefix` is better..

humblec commented:
> . So enableEpNamePrefix would sound more appropriate, though somewhat awkward -- maybe enableEpPrefix` is better..>
Reasonable justification :), renamed to `enableEpPrefix`.

humblec commented:
/test pull-kubernetes-bazel-test

jsafrane commented:
>> If you use PVC.UID as base of the name, is there any reason to make customepname configurable? It's introduced to solve the name length issue, but you don't have it now.

> Actually thats not the only reason. As the custom naming is an internal logic, I thought of providing an option to user to enable/disable.

What are the other reasons to introduce `customepname` (or `enableEpPrefix`)? Why would user want this thing configurable? IMO, it's internal implementation detail of the provisioner.

humblec commented:
>  
What are the other reasons to introduce customepname (or enableEpPrefix)? Why would user want this thing configurable? IMO, it's internal implementation detail of the provisioner.>
@jsafrane  The main idea behind this patch was a 2 phase approach:
1) GIve an option to admin to enable/disable , internal ep name prefix.
2) Allow admin to specify their own `prefix` string for ep. 

Thats why selection of `customepname` as the SC field/parameter came in first commit  and a follow up PR was targetted with `customepnameprefix` which can have any string  as per  admin's choice:

jsafrane commented:
> 1. GIve an option to admin to enable/disable , internal ep name prefix.
> 2. Allow admin to specify their own prefix string for ep.

Can you do both with the same `customepnameprefix` prefix? If it's set, use it, even if it's an empty string. If it's not set, use the default.

humblec commented:
> Can you do both with the same customepnameprefix prefix? If it's set, use it, even if it's an empty string. If it's not set, use the default.
>

@jsafrane  Done. Addressed in latest patch set.  PTAL . Thanks!

humblec commented:
/retest

humblec commented:
@jsafrane Please take a look at latest patch set whenever u have a free slot :). Thanks

jsafrane commented on pkg/volume/glusterfs/glusterfs_test.go:
```diff
@@ -645,6 +655,42 @@ func TestParseClassParameters(t *testing.T) {
+			},
+		},
+		{
+			"invalid customepnameprefix ( ie >26 chars) parameter",
```
I miss a test with empty prefix.

humblec replied:
Added in latest patchset.

jsafrane commented on pkg/volume/glusterfs/glusterfs_test.go:
```diff
@@ -645,6 +655,42 @@ func TestParseClassParameters(t *testing.T) {
+				"gidMin":             "4000",
+				"gidMax":             "5000",
+				"volumetype":         "replicate:4",
+				"customEpNamePrefix": "myprefixhasmorethan26characters",
```
Where is a check that the prefix is correctly cut at 26 characters?

humblec replied:
Added in latest patch set.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -1037,7 +1045,17 @@ func parseClassParameters(params map[string]string, kubeClient clientset.Interfa
+			} else {
+				return nil, fmt.Errorf("'customepnameprefix' value should be < %d characters", maxCustomEpNamePrefixLen)
+			}
+			cfg.customEpNamePrefix = v
```
The same line is already on line 1054  (in `if` branch).

humblec replied:
ah. removed.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -822,11 +826,14 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsVolum
 	// of volume creation.
-	epServiceName := dynamicEpSvcPrefix + p.options.PVC.Name
+
+	epServiceName = p.provisionerConfig.customEpNamePrefix + "-" + string(p.options.PVC.UID)
```
If user specifies `customepnameprefix: ""` (i.e. empty), I'd expect that the name of the service would be just the UID, not `-UID`.

jsafrane commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -67,13 +67,14 @@ var _ volume.Deleter = &glusterfsVolumeDeleter{}
 	gciLinuxGlusterMountBinaryPath = "/sbin/mount.glusterfs"
 	defaultGidMin                  = 2000
 	defaultGidMax                  = math.MaxInt32
+	maxCustomEpNamePrefixLen       = 26
```
Please add a comment where this number comes from (max length of name - length of uuid?)

humblec replied:
Done.

jsafrane reviewed (commented):
The logic is OK now, just a few comments here and there.

humblec commented:
@jsafrane Thanks!! Addressed review comments. PTAL.

obnoxxx reviewed (changes requested):
I think this patch is doing too much. Or at least too much that is not mentioned in the commit message: It claims to introduce the new SC parameter for custom prefix. But it also (and even more importantly, imho) changes the customEpName to be built with the UID instead of the name. This is a change I would definitely introduce in a separate patch.

Also, the original motivation for making the EpNamePrefix optional (and configurable) was to make it possible to have longer PVC names. But the same is already achieved by using the UID instead of the name to build the EpName. So if we put the change to use UID in first (maybe as a separate PR), do we need this one any more at all?

obnoxxx commented:
> > I think this patch is doing too much. Or at least too much that is not mentioned in the commit message: It claims to introduce the new SC parameter for custom prefix. But it also (and even more importantly, imho) changes the customEpName to be built with the UID instead of the name. This is a change I would definitely introduce in a separate patch.
> 
> Sure, I can seperate the commits for these changes. Or can update the commit message with some more details. Please let me know the preference.

If we agree that we want to **always** use the UID instead of the name for building the EpName (and I think we do?!...), then I would strongly prefer a separate patch (and even a separate PR) for this change to make it really clean and obvious what's going on.

(The only downside of always using UID that I see is that it might be a little more effort to align the EP with the PVC, right?)

> > Also, the original motivation for making the EpNamePrefix optional (and configurable) was to make it possible to have longer PVC names. 
> 
> Not really, this provides an option for admin to have their desired string as ep prefix.The advantage is that, its easy filering in a setup where many PVCs exist which was one of the user asks in the past.

That makes sense. It gives the admin the option to specify a custom prefix for each SC that uses this provisioner. This is nice. :+1:


### So let me summarize:

1. Based on the assumption that we want to change to always using the UID, I would suggest to have this as a separate patch, and if possible, separate PR. 
2. The introduction of the new parameter to control the EpNamePrefix can stay in this PR and can be rebased on the other.

humblec commented:
> Based on the assumption that we want to change to always using the UID, I would suggest to have this as a separate patch, and if possible, separate PR.>

Sure, here it is https://github.com/kubernetes/kubernetes/pull/70286

> The introduction of the new parameter to control the EpNamePrefix can stay in this PR and can be rebased on the other.>

Yep, I will rebase as soon as first PR got merged.

obnoxxx commented:
Thanks @humblec !

humblec committed:
glusterfs: Allow admin to provide  custom endpoint/service name prefix

This patch introduces a new SC parameter called `customepnameprefix`
in glusterfs plugin. This new storageclass parameter allow admins
to create endpoints and services with mentioned prefix.
If this parameter has not been set or if its empty,
default prefix string ie "glusterfs-dynamic-" will be used for ep/svc
name.

This patch address below issues#

https://github.com/kubernetes/kubernetes/issues/53939
https://github.com/kubernetes/kubernetes/issues/56379

Additional Ref# #
https://github.com/kubernetes/kubernetes/issues/56380

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
@jsafrane Rebased to master, Please review :)

humblec commented:
@jsafrane PTAL :) .. Thanks !!

jsafrane commented:
/lgtm

jsafrane commented:
/lgtm
/approve
